File: mns-reactor/src/test/java/cz/fb/manaus/reactor/price/PriceBulldozerTest.java
package cz.fb.manaus.reactor.price;

import cz.fb.manaus.core.model.Price;
import cz.fb.manaus.core.model.PriceComparator;
import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.model.TradedVolume;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Arrays;
import java.util.List;

import static java.util.Collections.singletonList;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThat;

public class PriceBulldozerTest extends AbstractLocalTestCase {

    @Autowired
    private PriceBulldozer bulldozer;

    private List<Price> SAMPLE = Arrays.asList(
            new Price(6d, 2d, Side.BACK),
            new Price(3d, 4d, Side.BACK),
            new Price(2.8, 10d, Side.BACK));

    @Test
    public void testBulldozeSimple() throws Exception {
        checkResult(3, Arrays.asList(new Price(3, 2, Side.LAY), new Price(4, 2, Side.LAY)), 1, 3.5, 4);
        List<Price> three = Arrays.asList(new Price(3, 2, Side.LAY),
                new Price(4, 2, Side.LAY), new Price(5, 2, Side.LAY));
        checkResult(3, three, 2, 3.5, 4);
        checkResult(5, three, 1, 4, 6);
    }

    @Test
    public void testBulldozeSingle() throws Exception {
        checkResult(3, singletonList(new Price(3, 2, Side.LAY)), 1, 3, 2);
        checkResult(1, singletonList(new Price(3, 2, Side.LAY)), 1, 3, 2);
    }

    @Test
    public void testBulldozeBoundary() throws Exception {
        List<Price> two = Arrays.asList(new Price(4, 2, Side.BACK), new Price(3, 2, Side.BACK));
        List<Price> three = Arrays.asList(new Price(5, 2, Side.BACK), new Price(4, 2, Side.BACK),
                new Price(3, 2, Side.BACK));

        checkResult(2, two, 2, 4, 2);
        checkResult(4, three, 2, 4.5, 4);
    }

    @Test
    public void testReal() throws Exception {
        checkResult(1, SAMPLE, 3, 6, 2);
        checkResult(2, SAMPLE, 3, 6, 2);
        checkResult(3, SAMPLE, 2, 4, 6);
        checkResult(4, SAMPLE, 2, 4, 6);
        checkResult(5, SAMPLE, 2, 4, 6);
        checkResult(6, SAMPLE, 2, 4, 6);
        checkResult(50, SAMPLE, 1, 3.25, 16);
        checkResult(100, SAMPLE, 1, 3.25, 16);
    }

    @Test(expected = IllegalStateException.class)
    public void testBadOrderBack() throws Exception {
        List<Price> badOrder = PriceComparator.ORDERING.reverse().sortedCopy(SAMPLE);
        bulldozer.bulldoze(10, badOrder);
    }

    @Test(expected = IllegalStateException.class)
    public void testBadOrderLay() throws Exception {
        bulldozer.bulldoze(10, Arrays.asList(new Price(5d, 2, Side.LAY), new Price(4d, 2, Side.LAY)));
    }

    private void checkResult(double threshold, List<Price> prices, int expectedCount,
                             double expectedPrice, double expectedAmount) {
        List<Price> bulldozed = bulldozer.bulldoze(threshold, prices);
        assertThat(bulldozed.size(), is(expectedCount));
        assertEquals(expectedPrice, bulldozed.get(0).getPrice(), 0.0001);
        assertEquals(expectedAmount, bulldozed.get(0).getAmount(), 0.0001);

        assertEquals(prices.stream().mapToDouble(Price::getAmount).sum(),
                bulldozed.stream().mapToDouble(Price::getAmount).sum(), 0.0001);
        assertEquals(TradedVolume.getWeightedMean(prices).getAsDouble(),
                TradedVolume.getWeightedMean(bulldozed).getAsDouble(), 0.0001);
    }

}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/price/PriceBulldozerTest.kt
package cz.fb.manaus.reactor.price

import cz.fb.manaus.core.model.Price
import cz.fb.manaus.core.model.PriceComparator
import cz.fb.manaus.core.model.Side
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class PriceBulldozerTest {

    private val realSample = listOf(
        Price(6.0, 2.0, Side.BACK),
        Price(3.0, 4.0, Side.BACK),
        Price(2.8, 10.0, Side.BACK)
    )

    @Test
    fun `bulldoze lay prices`() {
        checkResult(3.0, listOf(Price(3.0, 2.0, Side.LAY), Price(4.0, 2.0, Side.LAY)), 1, 3.5, 4.0)
        val three = listOf(
            Price(3.0, 2.0, Side.LAY),
            Price(4.0, 2.0, Side.LAY), Price(5.0, 2.0, Side.LAY)
        )
        checkResult(3.0, three, 2, 3.5, 4.0)
        checkResult(5.0, three, 1, 4.0, 6.0)
    }

    @Test
    fun `bulldoze - only 1 price`() {
        checkResult(3.0, listOf(Price(3.0, 2.0, Side.LAY)), 1, 3.0, 2.0)
        checkResult(1.0, listOf(Price(3.0, 2.0, Side.LAY)), 1, 3.0, 2.0)
    }

    @Test
    fun `threshold equal to amount sum`() {
        val two = listOf(Price(4.0, 2.0, Side.BACK), Price(3.0, 2.0, Side.BACK))
        val three = listOf(
            Price(5.0, 2.0, Side.BACK), Price(4.0, 2.0, Side.BACK),
            Price(3.0, 2.0, Side.BACK)
        )

        checkResult(2.0, two, 2, 4.0, 2.0)
        checkResult(4.0, three, 2, 4.5, 4.0)
    }

    @Test
    fun `bulldoze real data`() {
        checkResult(1.0, realSample, 3, 6.0, 2.0)
        checkResult(2.0, realSample, 3, 6.0, 2.0)
        checkResult(3.0, realSample, 2, 4.0, 6.0)
        checkResult(4.0, realSample, 2, 4.0, 6.0)
        checkResult(5.0, realSample, 2, 4.0, 6.0)
        checkResult(6.0, realSample, 2, 4.0, 6.0)
        checkResult(50.0, realSample, 1, 3.25, 16.0)
        checkResult(100.0, realSample, 1, 3.25, 16.0)
    }

    fun `bulldozed prices are in wrong order - back`() {
        assertFailsWith<IllegalStateException> {
            val badOrder = realSample.sortedWith(PriceComparator).reversed()
            PriceBulldozer.bulldoze(10.0, badOrder)
        }
    }

    @Test
    fun `bulldozed prices are in wrong order - lay`() {
        assertFailsWith<IllegalStateException> {
            PriceBulldozer.bulldoze(
                10.0,
                listOf(Price(5.0, 2.0, Side.LAY), Price(4.0, 2.0, Side.LAY))
            )

        }
    }

    private fun checkResult(
        threshold: Double,
        prices: List<Price>,
        expectedCount: Int,
        expectedPrice: Double,
        expectedAmount: Double
    ) {
        val bulldozed = PriceBulldozer.bulldoze(threshold, prices)
        assertEquals(expectedCount, bulldozed.size)
        assertEquals(expectedPrice, bulldozed[0].price, 0.0001)
        assertEquals(expectedAmount, bulldozed[0].amount, 0.0001)

        assertEquals(
            prices.sumOf { it.amount },
            bulldozed.sumOf { it.amount }, 0.0001
        )
        assertEquals(
            getWeightedMean(prices)!!,
            getWeightedMean(bulldozed)!!, 0.0001
        )
    }

}